### PR TITLE
feat: use real launch templates in validation instead of dry-run

### PR DESCRIPTION
**Pull request by ryan-mist**
Fixes #N/A

**Description**
* replace dry run validation of `CreateLaunchTemplate` with creating a real launch template
* utilized real launch template in validation for `CreateFleet` and `RunInstances`

**How was this change tested?**
* `make presubmit`
* `/karpenter snapshot`
* manual testing by running binary against cluster

**Test Case 1** - Valid Config
* ensuring Karpenter has necessary permissions to pass validation
* ensuring validation controller (1) creates launch template, (2) passes all validation checks, and (3) updates node class validation status to show it succeeded
Screenshot 2025-08-21 at 3 43 48 PM

**Test Case 2** - Invalid Config due to `CreateFleet` Setup
* ensuring Karpenter has necessary permissions to pass validation except for `CreateFleet` validation
  * i.e. adding deny policy for `ec2:CreateFleet`
* ensuring validation controller (1) creates launch template, (2) fails create fleet validation check, and (3) updates node class validation status to show it failed
Screenshot 2025-08-21 at 3 47 08 PM

**Test Case 3** - Invalid Config due to `CreateLaunchTemplate` Setup
* ensuring Karpenter has necessary permissions to pass validation except for `CreateLaunchTemplate` validation
  * i.e. adding deny policy for `ec2:CreateLaunchTemplate`
* ensuring validation controller (1) fails to create launch template and (2) updates node class validation status to show it failed
Screenshot 2025-08-21 at 3 48 48 PM

**Test Case 4** - Invalid Config due to `RunInstances` Setup
* ensuring Karpenter has necessary permissions to pass validation except for `RunInstances` validation
  * i.e. adding deny policy for `ec2:RunInstances`
* ensuring validation controller (1) creates launch template, (2) fails run instances validation check, and (3) updates node class validation status to show it failed
Screenshot 2025-08-21 at 4 05 34 PM

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/nodeclass/controller.go [hunk 1]
```diff
@@ -104,8 +105,8 @@ func NewController(
 			NewSubnetReconciler(subnetProvider),
 			NewSecurityGroupReconciler(securityGroupProvider),
 			NewInstanceProfileReconciler(instanceProfileProvider, region, recreationCache),
-			validation,
 			NewReadinessReconciler(launchTemplateProvider),
+			validation,
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/controller.go:107]: Is there a way to build that logic into the validation controller directly so that we aren't relying on implicit ordering here to ensure that things don't break for folks that are using AL2023 and need to have resolved CIDRs?

File: pkg/controllers/nodeclass/validation.go [hunk 2]
```diff
@@ -183,47 +201,63 @@ func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass)
...
-		}
-		return ConditionReasonCreateFleetAuthFailed, false, nil
+		return ConditionReasonCreateLaunchTemplateAuthFailed, false, nil
+	}
+	// this case should never occur
+	if len(launchTemplates) == 0 {
+		return "", false, fmt.Errorf("no launch templates created")
 	}
+	// update validation context
+	runInstancesInput := &ec2.RunInstancesInput{}
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go]: Did we validate if we can just pass the launch template to RunInstances and have it valid all of the details of the launch template for us? This would prevent us from having to generate all the RunInstances input. We avoided this in the past because we wanted to avoid the launch template creation, but now, that seems inevitable

File: pkg/controllers/nodeclass/validation.go [hunk 3]
```diff
@@ -183,47 +201,63 @@ func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass)
...
+		return "", false, fmt.Errorf("generating options, %w", err)
+	}
+	// this case should never occur
+	if len(opts.InstanceTypes) == 0 {
+		return "", false, fmt.Errorf("no instance types available")
+	}
+	// we only want to create 1 launch template so we only pass 1 instance type to EnsureAll
+	instanceTypes := opts.InstanceTypes[:1]
+	launchTemplates, err := v.launchTemplateProvider.EnsureAll(ctx, nodeClass, nodeClaim, instanceTypes, karpv1.CapacityTypeOnDemand, tags)
+	if err != nil {
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go:216]: Authorization isn't the only reason that this would fail, right? Should we be more precise with how we handle failures in launch template creation here so that we can identify the difference between a retryable error and an authorization error

File: pkg/controllers/nodeclass/controller.go (hunk 1 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/controller.go:107]: We could copy the logic from the readiness controller into the validation controller.
i.e. call something similar to the following before running our dry-runs
```
if nodeClass.AMIFamily() == v1.AMIFamilyAL2023 {
	if err := n.launchTemplateProvider.ResolveClusterCIDR(ctx); err != nil {
		nodeClass.StatusConditions().SetFalse(status.ConditionReady, "NodeClassNotReady", "Failed to detect the cluster CIDR")
		return reconcile.Result{}, fmt.Errorf("failed to detect the cluster CIDR, %w", err)
	}
}
```
`ResolveClusterCIDR` checks to see if the CIDR has already been loaded so there wouldn't be any additional API calls

File: pkg/controllers/nodeclass/validation.go (hunk 2 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go]: Good point, will validate now

File: pkg/controllers/nodeclass/validation.go (hunk 3 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go:216]: good point, will add more precise error handling

File: pkg/controllers/nodeclass/validation.go (hunk 2 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go]: validated this and updated call

File: pkg/controllers/nodeclass/controller.go (hunk 1 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/controller.go:107]: made change

File: pkg/controllers/nodeclass/validation.go (hunk 3 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go:216]: added extra error handling for server related errors that would imply a quick retry is needed

**Review by ryan-mist — COMMENTED** [review]
/karpenter snapshot

File: pkg/controllers/nodeclass/controller.go (hunk 1 above)
Comment (jonathan-innis) [reply pkg/controllers/nodeclass/controller.go:107]: In general, nothing should be setting the readiness condition directly like the cluster CIDR resolution is doing -- can we just group this into the validation controller like you are doing now and just drop the readiness sub-reconciler outright

File: pkg/controllers/nodeclass/validation.go [hunk 4]
```diff
@@ -120,6 +125,25 @@ func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass)
 		)
 		return reconcile.Result{}, nil
 	}
+	// Ensure CIDR is resolved for relevant AMIs, if it is not we know validation will fail regardless of the other values.
+	// We first check ready condition to not re-call API
+	readyCondition := nodeClass.StatusConditions().Get(status.ConditionReady)
+	if readyCondition.Reason == NodeClassNotReady && readyCondition.Message == CIDRDetectionFailed {
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go]: I don't think that we need both of these -- let's just do the CIDR resolution here if we have an AL2023 AMI -- that should basically give us what we have today but ensure that we aren't messing with the readiness condition of the nodeclass directly

File: pkg/controllers/nodeclass/controller.go (hunk 1 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/controller.go:107]: Yes, will do (edit: done)

File: pkg/controllers/nodeclass/validation.go [hunk 5]
```diff
@@ -157,13 +181,14 @@ func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass)
 		v.cache.SetDefault(v.cacheKey(nodeClass, tags), "")
 		return reconcile.Result{}, nil
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go]: QQ: Is this line correct? We don't requeue from this controller to re-check our validation? That seems wrong to me -- we had promised that we would do this on a poll, but we aren't requiring that we do this on a poll in this controller

File: pkg/controllers/nodeclass/validation.go [hunk 6]
```diff
@@ -157,13 +181,14 @@ func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass)
...
 	}
+	validationCtx := &validationContext{}
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go]: nit: We have a check below where we are setting our status to True if dry-run is disabled, I'm not following why we are setting a reason associated with this -- we do other validations outside of dry-run that should still be relevant so I'm not following why we have a very specific reason here

File: pkg/controllers/nodeclass/validation.go [hunk 7]
```diff
@@ -183,47 +208,54 @@ func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass)
 	return reconcile.Result{}, nil
 }
 
-type validatorFunc func(context.Context, *v1.EC2NodeClass, *karpv1.NodeClaim, map[string]string) (string, bool, error)
+type validatorFunc func(context.Context, *v1.EC2NodeClass, *karpv1.NodeClaim, map[string]string, *validationContext) (string, bool, error)
 
-func (v *Validation) validateCreateFleetAuthorization(
+func (v *Validation) validateCreateLaunchTemplateAuthorization(
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go:204]: Mutating a context that is passed in like this and passing this to the next function is a little bit of an odd pattern and hard to follow. In general, I think we are somewhat pigeon-holing our design into the `isValid()` "ValidatorFunc" paradigm. If we aren't required to fulfill this interface and unravel some of these things, does it make it easier to pass the launch templates around to the validator functions?

File: pkg/controllers/nodeclass/validation.go (hunk 4 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go]: moved readiness logic into validation controller

File: pkg/controllers/nodeclass/validation.go [hunk 8]
```diff
@@ -183,47 +198,54 @@ func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass)
...
+	validationCtx *validationContext,
 ) (reason string, requeue bool, err error) {
-	createFleetInput := instance.NewCreateFleetInputBuilder(karpv1.CapacityTypeOnDemand, tags, mockLaunchTemplateConfig()).Build()
-	createFleetInput.DryRun = lo.ToPtr(true)
-	// Adding NopRetryer to avoid aggressive retry when rate limited
-	if _, err := v.ec2api.CreateFleet(ctx, createFleetInput, func(o *ec2.Options) {
-		o.Retryer = aws.NopRetryer{}
-	}); awserrors.IgnoreDryRunError(err) != nil {
-		if awserrors.IsRateLimitedError(err) {
+	opts, err := v.getLaunchTemplateOptions(ctx, nodeClaim, nodeClass, tags)
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go]: I'm still not following why we need to get the launch template options at all like this if we don't use this anywhere and EnsureAll just generates the launch template for us

**Review by ryan-mist — COMMENTED** [review]
/karpenter snapshot

File: pkg/controllers/nodeclass/validation.go (hunk 5 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go]: In the validation we always return `reconcile.Result{}`. My best guess as to why it because in the main nodeclass controller we take the minimum over all the reconcile results from nodeclass controllers.  We're guaranteed to have a retry time no longer than 5 min due to the nodeclass AMI controller (unless there is a error in which case we return `reconcile.Result{}` in the nodeclass controller anyway)

File: pkg/controllers/nodeclass/validation.go (hunk 6 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go:174]: Maybe a reason along the lines of "Required conditions met and dry run is disabled" to align with the other validations outside of dry-run?

File: pkg/controllers/nodeclass/validation.go (hunk 8 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go:210]: This function gets the instanceTypes to pass into EnsureAll. Since this is the only use of the function, I'll scope down the function to whats important.

File: pkg/controllers/nodeclass/validation.go (hunk 7 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go:203]: We could but `validateCreateLaunchTemplateAuthorization` is a `validatorFunc`, so it makes a bit tougher to move the created launch template around in a clean way as we create it in that function. 

If we return launch template from `validateCreateLaunchTemplateAuthorization` then we would have to change the return type of `validatorFunc` (or just make it so `validateCreateLaunchTemplateAuthorization` is not a `validatorFunc` which also seems off). Another way is just passing a pointer in and modifying it in `validateCreateLaunchTemplateAuthorization` but that is essentially what validation context is.

File: pkg/controllers/nodeclass/validation.go (hunk 5 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go]: I can add reconcile.Result{10*time.Minute} to remove this implicit dependency

File: pkg/controllers/nodeclass/validation.go (hunk 8 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go]: We can't replace this function with just a call to something like `getInstanceTypesForNodeClass` because we prioritize certain instance types which will launch with a non-GPU AMI due to this issue - https://github.com/aws/karpenter-provider-aws/issues/7928. A call to `EnsureAll` with the list of instances from `getInstanceTypesForNodeClass` will not do the same (and subsequently fail our unit tests)

File: pkg/controllers/nodeclass/validation.go (hunk 7 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go:204]: Yes, removed validation func paradigm and just unrolled.

File: pkg/controllers/nodeclass/validation.go (hunk 6 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go]: Changed to SetTrue rather than SetTrueWithReason

File: pkg/controllers/nodeclass/validation.go (hunk 8 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go]: How do you feel about replacing this call to `EnsureAll` with `ensureLaunchTemplate` (and making it public)? This reduces the duplicated work in resolving the launch template in `EnsureAll` as we already resolved it the validation controller

File: pkg/controllers/nodeclass/validation.go [hunk 9]
```diff
@@ -120,6 +120,15 @@ func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass)
 		)
 		return reconcile.Result{}, nil
 	}
+	// A NodeClass that uses AL2023 requires the cluster CIDR for launching nodes.
+	// To allow Karpenter to be used for Non-EKS clusters, resolving the Cluster CIDR
+	// will not be done at startup but instead in a reconcile loop.
+	if nodeClass.AMIFamily() == v1.AMIFamilyAL2023 {
+		if err := v.launchTemplateProvider.ResolveClusterCIDR(ctx); err != nil {
+			nodeClass.StatusConditions().SetFalse(status.ConditionReady, "NodeClassNotReady", "Failed to detect the cluster CIDR")
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go]:
```suggestion
			nodeClass.StatusConditions().SetFalse(status.ConditionTypeValidationSucceeded, "ClusterCIDRResolutionFailed", "Failed to detect the cluster CIDR")
```

We should potentially also consider how we can keep this in an Unknown state until we get a known failure rather than throwing this to False when we have a transient issue

File: pkg/controllers/nodeclass/validation.go (hunk 5 above)
Comment (jonathan-innis) [reply pkg/controllers/nodeclass/validation.go]: Yep, I knew that that was there, but I'd still like these sub-reconcilers to be self-contained to avoid introducing weird bugs later

File: pkg/errors/errors.go [hunk 10]
```diff
@@ -150,6 +152,23 @@ func IgnoreRateLimitedError(err error) error {
 	return err
 }
 
+func IsTemporaryServiceError(err error) bool {
```
Comment (jonathan-innis) [inline pkg/errors/errors.go]: Should this be something like "IsRetryable"? Does the AWS SDK natively support a function that checks for retryable errors like this?

File: pkg/controllers/nodeclass/validation.go [hunk 11]
```diff
@@ -97,6 +97,16 @@ func NewValidationReconciler(
 
 // nolint:gocyclo
 func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass) (reconcile.Result, error) {
+	// A NodeClass that uses AL2023 requires the cluster CIDR for launching nodes.
+	// To allow Karpenter to be used for Non-EKS clusters, resolving the Cluster CIDR
+	// will not be done at startup but instead in a reconcile loop.
+	if nodeClass.AMIFamily() == v1.AMIFamilyAL2023 {
+		if err := v.launchTemplateProvider.ResolveClusterCIDR(ctx); err != nil {
+			nodeClass.StatusConditions().SetFalse(status.ConditionReady, "NodeClassNotReady", "Failed to detect the cluster CIDR")
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go]:
```suggestion
			nodeClass.StatusConditions().SetFalse(status. ConditionTypeValidationSucceeded, "ClusterCIDRResolutionFailed", "Failed to detect the cluster CIDR")
```

We still shouldn't be directly setting the Ready status condition here

File: pkg/controllers/nodeclass/validation.go [hunk 12]
```diff
@@ -97,6 +97,16 @@ func NewValidationReconciler(
...
+			return reconcile.Result{}, fmt.Errorf("failed to detect the cluster CIDR, %w", err)
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go:114]: Similar to what we're talking about doing elsewhere, is there a way to differentiate between transient and non-transient errors here to know when to set it to Unknown or False

File: pkg/controllers/nodeclass/validation.go (hunk 8 above)
Comment (jonathan-innis) [reply pkg/controllers/nodeclass/validation.go]: > because we prioritize certain instance types which will launch with a non-GPU AMI due to this issue

I'm confused why we can't just pass-in the correct instance types based on this prioritized set of requirements. I might be missing something, though, so let's discuss if that still doesn't make sense

File: pkg/controllers/nodeclass/validation.go [hunk 13]
```diff
@@ -335,29 +332,71 @@ func (v *Validation) clearCacheEntries(nodeClass *v1.EC2NodeClass) {
 	}
 }
 
-func mockLaunchTemplateConfig() []ec2types.FleetLaunchTemplateConfigRequest {
+func getRunInstancesInput(
+	nodeClass *v1.EC2NodeClass,
+	tags map[string]string,
+	launchTemplate *launchtemplate.LaunchTemplate,
+) *ec2.RunInstancesInput {
+	return &ec2.RunInstancesInput{
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go:333]: If we pass-in a launch template here, doesn't that cover the networkInterface specification, imageID, instance type, etc. or we still have to pass these in manually since these aren't included in the launch template since we assumed that they would be passed-through in the CreateFleet request?

File: pkg/controllers/nodeclass/validation.go (hunk 8 above)
Comment (jonathan-innis) [reply pkg/controllers/nodeclass/validation.go]: Maybe the confusion here is that I don't think we need to return the full set of launch template options, we just need the instance types that we are going to pass around to our functions, so maybe we should just change the naming?

File: pkg/controllers/nodeclass/validation.go [hunk 14]
```diff
@@ -149,81 +159,97 @@ func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass)
...
 	v.cache.SetDefault(v.cacheKey(nodeClass, tags), "")
 	nodeClass.StatusConditions().SetTrue(v1.ConditionTypeValidationSucceeded)
 	return reconcile.Result{}, nil
 }
 
-type validatorFunc func(context.Context, *v1.EC2NodeClass, *karpv1.NodeClaim, map[string]string) (string, bool, error)
-
-func (v *Validation) validateCreateFleetAuthorization(
-	ctx context.Context,
+func (v *Validation) handleValidationResult(
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go]: This function should also requeueAfter: 10m if we are planning to requeue consistently -- we might be able to factor this slightly different if we return RequeueAfter, actually. Returning a return argument like `shouldReturn` feels a little off. What if we can check if the result is empty? If it's empty, then we just move on to the next function -- the only ways to return early for this function are to set a RequeueAfter or return an error?

File: pkg/controllers/nodeclass/validation_test.go [hunk 15]
```diff
@@ -30,6 +30,7 @@ import (
 	"k8s.io/client-go/tools/record"
 	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
 	"sigs.k8s.io/karpenter/pkg/events"
+	coreoptions "sigs.k8s.io/karpenter/pkg/operator/options"
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation_test.go]: Are there no other changes to our validation tests that need to be made by changing what we are passing into the CreateFleet and RunInstances calls?

File: pkg/controllers/nodeclass/validation.go (hunk 5 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go]: Sounds good, ill change it then

File: pkg/controllers/nodeclass/validation.go (hunk 13 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go:333]: It does not require `imageID` and `instanceType` but I added them to be compatible with current tests as it doesn't change the validation of the call. I meant to change tests to not require it, but Ill change that now.

For `networkInterface` specification it does require (currently) it as the subnets weren't built into the spec on launch template creation. I can look into why the launch template is not generated with to attempt to fix that.

File: pkg/controllers/nodeclass/validation.go (hunk 8 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go]: Took another run at trying to get this working. My guess is we only want 1 launch template to be created in validation. We can prioritize the instance types in `validation.go` but can't naively pass in the first instance type (to ensure 1 launch template creation) in the list of prioritizations as `amiFamily.Resolve` does additional prioritizations which requires the full list of instance types. 

Before, we were only taking the first item from the return value of `amiFamily.Resolve`. We can replicate this functionality in the call to `EnsureAll` but, we only want to do this in the case of validation. How do you feel about adding in an extra argument to `EnsureAll` (i.e. a boolean value `createOneLaunchTemplate`) which does this?

File: pkg/controllers/nodeclass/validation_test.go (hunk 15 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation_test.go]: I don't believe so. Existing tests already validate that when an API call is unauthorized, then the relevant status condition on the node class is shown.

e.g.
```
Entry("should update status condition as NotReady when CreateFleet unauthorized", func() {
	awsEnv.EC2API.CreateFleetBehavior.Error.Set(&smithy.GenericAPIError{
		Code: "UnauthorizedOperation",
	}, fake.MaxCalls(1))
}, nodeclass.ConditionReasonCreateFleetAuthFailed),

Entry("should update status condition as NotReady when RunInstances unauthorized", func() {
	awsEnv.EC2API.RunInstancesBehavior.Error.Set(&smithy.GenericAPIError{
		Code: "UnauthorizedOperation",
	}, fake.MaxCalls(1))
}, nodeclass.ConditionReasonRunInstancesAuthFailed),

Entry("should update status condition as NotReady when CreateLaunchTemplate unauthorized", func() {
	awsEnv.EC2API.CreateLaunchTemplateBehavior.Error.Set(&smithy.GenericAPIError{
		Code: "UnauthorizedOperation",
	}, fake.MaxCalls(1))
}, nodeclass.ConditionReasonCreateLaunchTemplateAuthFailed),
```

The reason launchtemplate_test.go needed to be updated to mock an API call is because they define the nodeclass differently there. Another test case I could add is to make sure if we create a launch template then run validation again we don't call create launch template (rather use the one in the cache). I don't think its exactly needed as thats more of a launch template functionality.

File: pkg/controllers/nodeclass/validation.go (hunk 13 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go:333]: * For `imageID`, we can rely on launch template to have this data, will remove this from run instances input.
* For `instanceType`, the launch template is not built with this data so will keep here (won't impact verification as that is the corresponding instance type for the launch template anyway).
* From what I've found for `networkInterface`, network interfaces for the launch template are not guaranteed to be generated if the instance does not define `Capacity[v1.ResourceEFA]`. If we don't define this in the run instances call and the launch template does not define a network interface we get `InvalidParameterCombination`.

File: pkg/errors/errors.go (hunk 10 above)
Comment (ryan-mist) [reply pkg/errors/errors.go]: Good point, we can check if the `ErrorFault` is `FaultServer`

File: pkg/controllers/nodeclass/validation.go (hunk 12 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go:114]: Will add the same `awserrors.IsServerError(err)` check

File: pkg/controllers/nodeclass/validation.go (hunk 8 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go]: Changed to use `getPrioritizedInstanceTypes`

**Review by ryan-mist — COMMENTED** [review]
/karpenter snapshot

File: pkg/controllers/nodeclass/launchtemplate_test.go [hunk 16]
```diff
@@ -75,7 +88,7 @@ var _ = Describe("NodeClass Launch Template CIDR Resolution Controller", func()
 		ExpectObjectReconciled(ctx, env.Client, controller, nodeClass)
 		Expect(lo.FromPtr(awsEnv.LaunchTemplateProvider.ClusterCIDR.Load())).To(Equal("10.100.0.0/16"))
 		nodeClass = ExpectExists(ctx, env.Client, nodeClass)
-		Expect(nodeClass.StatusConditions().IsTrue(status.ConditionReady)).To(BeTrue())
+		Expect(nodeClass.StatusConditions().IsTrue(v1.ConditionTypeValidationSucceeded)).To(BeTrue())
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/launchtemplate_test.go]: Can we just move these tests out of the launch template provider test suite? They don't really belong here since that reconciler is being run elsewhere -- I'd rather just have them in the nodeclass suite

File: pkg/controllers/nodeclass/validation.go [hunk 17]
```diff
@@ -97,6 +98,28 @@ func NewValidationReconciler(
 
 // nolint:gocyclo
 func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass) (reconcile.Result, error) {
+	// A NodeClass that uses AL2023 requires the cluster CIDR for launching nodes.
+	// To allow Karpenter to be used for Non-EKS clusters, resolving the Cluster CIDR
+	// will not be done at startup but instead in a reconcile loop.
+	if nodeClass.AMIFamily() == v1.AMIFamilyAL2023 {
+		if err := v.launchTemplateProvider.ResolveClusterCIDR(ctx); err != nil {
+			if awserrors.IsServerError(err) {
+				nodeClass.StatusConditions().SetUnknownWithReason(
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go]: I think similar to how we just requeue on server errors below, we can probably leave the status condition and requeue here while only making sure that we set false when we have a real error

**Review by ryan-mist — COMMENTED** [review]
/karpenter snapshot

File: pkg/controllers/nodeclass/validation.go (hunk 13 above)
Comment (jonathan-innis) [reply pkg/controllers/nodeclass/validation.go:333]: Got it -- so we basically have to define a network interface here, makes sense

File: pkg/controllers/nodeclass/validation.go [hunk 18]
```diff
@@ -409,13 +406,13 @@ func (v *Validation) mockLaunchTemplateOptions(
 	// wouldn't be chosen due to cost in practice. This ensures the behavior matches that on Karpenter v1.3, preventing a
 	// potential regression for Windows users.
 	// Tracking issue: https://github.com/aws/karpenter-provider-aws/issues/7985
-	if len(selectedInstanceTypes) == 0 || lo.ContainsBy([]string{
+	if len(instanceTypes) == 0 || lo.ContainsBy([]string{
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go:409]: Doesn't re-ordering this check change the behavior? Before, you were only picking these default instance types if there was no AMI that mapped to the matching instance types; now, you are just doing it if you don't get compatible instance types back generally

File: pkg/controllers/nodeclass/validation.go [hunk 19]
```diff
@@ -145,154 +168,126 @@ func (v *Validation) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass)
...
+	}
+	// no instance types are compatible with the given AMI, so validation we know fail
+	if len(instanceTypes) == 0 {
+		return nil, reconcile.Result{}, fmt.Errorf("no instance types available for AMI")
+	}
+	// pass 1 instance type in EnsureAll to only create 1 launch template
+	launchTemplates, err := v.launchTemplateProvider.EnsureAll(ctx, nodeClass, nodeClaim, instanceTypes[:1], karpv1.CapacityTypeOnDemand, tags)
+	if err != nil {
+		if awserrors.IsRateLimitedError(err) || awserrors.IsServerError(err) {
+			return nil, reconcile.Result{Requeue: true}, nil
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation.go:218]: This doesn't check for authorization like the other methods do -- should it?

File: pkg/controllers/nodeclass/validation_test.go [hunk 20]
```diff
@@ -181,7 +232,7 @@ var _ = Describe("NodeClass Validation Status Controller", func() {
 				}, fake.MaxCalls(1))
 			}, nodeclass.ConditionReasonCreateLaunchTemplateAuthFailed),
 		)
-		Context("RunInstances Validation", func() {
+		Context("Instance Type Prioritization Validation", func() {
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation_test.go:264]: Why did this naming change?

File: pkg/controllers/nodeclass/validation_test.go [hunk 21]
```diff
@@ -88,6 +89,56 @@ var _ = Describe("NodeClass Validation Status Controller", func() {
 			Entry(v1.ConditionTypeSubnetsReady, v1.ConditionTypeSubnetsReady),
 		)
 	})
+	Context("NodeClass Status Conditions", func() {
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation_test.go]: This is a pretty generic title. Also, I'm curious what failure case this is meant to catch for reserved capacity

File: pkg/controllers/nodeclass/validation.go (hunk 19 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go:218]: Good point, added

File: pkg/controllers/nodeclass/validation.go (hunk 18 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation.go:409]: Added additional AMI mapping in getInstanceTypesForNodeClass to remove edge case.

File: pkg/controllers/nodeclass/validation_test.go (hunk 20 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation_test.go:264]: Changed naming because all test cases in this context dealt with prioritizing instance types. Since this prioritization is also a part of launch template creation (since we pass this into run instances now) it is now more general than just run instances validation.

File: pkg/controllers/nodeclass/launchtemplate_test.go (hunk 16 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/launchtemplate_test.go]: These tests are already in the nodeclass suite right? i.e. in controllers/nodeclass not providers/launchtemplate

File: pkg/controllers/nodeclass/validation_test.go (hunk 21 above)
Comment (ryan-mist) [reply pkg/controllers/nodeclass/validation_test.go]: These were originally in `pkg/controllers/nodeclass/readiness_test.go`. Since we moved the functionality of `pkg/controllers/nodeclass/readiness.go` into the validation controller I also removed this test file and moved it into here. Should I move back into `readiness_test.go`? It doesn't seem to test the CIDR functionality that `readiness.go` controller actually implemented.

Comment (jonathan-innis) [reply pkg/controllers/nodeclass/validation_test.go]: I think we should just move these tests into their relevant sub-reconcilers

File: pkg/controllers/nodeclass/launchtemplate_test.go (hunk 16 above)
Comment (jonathan-innis) [reply pkg/controllers/nodeclass/launchtemplate_test.go]: Ah, makes sense. Sounds good. You could argue that these should just be put into the validation_test.go since it has to do with things that are executed in the validation sub-reconciler

Comment (ryan-mist) [reply pkg/controllers/nodeclass/launchtemplate_test.go]: Makes sense Ill move them over

File: pkg/controllers/nodeclass/validation_test.go [hunk 22]
```diff
@@ -88,6 +91,83 @@ var _ = Describe("NodeClass Validation Status Controller", func() {
 			Entry(v1.ConditionTypeSubnetsReady, v1.ConditionTypeSubnetsReady),
 		)
 	})
+	Context("CIDR Resolution Precondion", func() {
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/validation_test.go]:
```suggestion
	Context("CIDR Resolution Precondition", func() {
```

😢

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀
